File: 5week/김수현/1260. DFS와 BFS/DFS와 BFS.js
const input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');

function dfs(graph, startNode) {
   const visited = [];
   let needVisit = [];

   needVisit.push(startNode);
   while(needVisit.length !==0){
      const node = needVisit.pop();
      if(!visited.includes(node)){
         visited.push(node);
         let nodes = graph[node];
         needVisit = [...needVisit, ...nodes.sort((a, b) => b - a)];
      }
   }

   return visited;
};

function bfs(graph, startNode){
   const visited = [];
   let needVisit = [];

   needVisit.push(startNode);

   while(needVisit.length !==0){
      const node = needVisit.shift();
      if(!visited.includes(node)){
         visited.push(node);
         let nodes = graph[node];
         needVisit = [...needVisit, ...nodes.sort((a, b) => a - b)];
      }
   }

   return visited;
}

let [n, edge, start] = input.shift().split(' ').map(Number);
let graph = [...Array(n + 1)].map(e => []);

for(let i = 0 ; i < edge ; i++){
   let [from, to] = input[i].split(' ').map(Number);
   graph[from].push(to);
   graph[to].push(from);
}

console.log(dfs(graph, start).join(' '));
console.log(bfs(graph, start).join(' '));

File: 5week/심예은/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.*;

public class Main {
    static ArrayList<Integer>[] adjList;
    static boolean[] visited;
    static StringBuilder dfsResult = new StringBuilder();
    static StringBuilder bfsResult = new StringBuilder();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        
        String[] input = br.readLine().split(" ");
        int N = Integer.parseInt(input[0]);
        int M = Integer.parseInt(input[1]);
        int V = Integer.parseInt(input[2]);

        adjList = new ArrayList[N + 1];
        for (int i = 1; i <= N; i++) {
            adjList[i] = new ArrayList<>();
        }

        for (int i = 0; i < M; i++) {
            input = br.readLine().split(" ");
            int a = Integer.parseInt(input[0]);
            int b = Integer.parseInt(input[1]);

            adjList[a].add(b);
            adjList[b].add(a);
        }

        for (int i = 1; i <= N; i++) {
            Collections.sort(adjList[i]);
        }

        visited = new boolean[N + 1];
        dfs(V);

        visited = new boolean[N + 1];
        bfs(V);

        System.out.println(dfsResult.toString().trim());
        System.out.println(bfsResult.toString().trim());
    }

    public static void dfs(int node) {
        if (visited[node]) return;

        visited[node] = true;
        dfsResult.append(node).append(" ");

        for (int nextNode : adjList[node]) {
            if (!visited[nextNode]) {
                dfs(nextNode);
            }
        }
    }

    public static void bfs(int start) {
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        visited[start] = true;

        while (!queue.isEmpty()) {
            int node = queue.poll();
            bfsResult.append(node).append(" ");

            for (int nextNode : adjList[node]) {
                if (!visited[nextNode]) {
                    visited[nextNode] = true;
                    queue.add(nextNode);
                }
            }
        }
    }
}
